### DISCUSSION TITLE
Compatibility with Dreame L40

### DISCUSSION THREAD

**Discussion by incline02** [sca075/mqtt_vacuum_camera#306] [Q&A] [answered]
### Description

As of now, I have my vacuum bot entity. When I try to create an MQTT Vacuum Camera integration, on the very first screen (Camera setup), I select my vacuum entity and click submit and get a "Unknown error occurred".  talking to some other users, I hear it's not compatible yet with Dreame L40

### Solution

compatibility with dreame L40

### Alternatives

I've tried HomeAssistant stock camera integration, homekit integration, Proxy camera, and go2rtC they all give the same results when I try to watch the live stream full screen in homekit: When watching the preview of camera, it seems to update every 5 seconds. When going into the live stream, it's on constant buffer.

**Comment by sca075** [comment, marked as answer]
@incline02 
This integration works with all vacuums integrated via mqtt:
![Screenshot 2024-12-13 at 18 42 52](https://github.com/user-attachments/assets/1125f8d6-5d67-4970-9397-40ad94c14de0)
![Screenshot 2024-12-13 at 18 39 49](https://github.com/user-attachments/assets/ab690e9b-ac28-412d-8393-486461ba70a9)
The error indicates that your vacuum somehow do not provide any data to MQTT, if you are using the stock firmware so you can use for Dreame L40 [this integration instead](https://github.com/Tasshack/dreame-vacuum). 
Sorry but can't do that within this project... unless.. the vacuum entity is providing this data to MQTT... I know it seams a little strange.. but soon I will release some library that can probably be used...

**Comment by sca075** [comment]
@incline02  I did read the conversation you had [here](https://community.home-assistant.io/t/valetudo-map-card-installation-instructions/247329/28) I'm wander is the vacuum is having a camera entity with data in "valetudo" format.. probably.. but this is a huge work around.. if the vacuum data instead are provided as an MQTT entity then this camera could work.

**Reply by incline02** [reply]
Thank you so much for the replies.

I understand and wouldn't have you pull that kind of legwork, I only thought compatibility. Can I ask what alternative do you suggest is best? Should it be the maintainer of the dreame integration instead or is it just the robot that's not providing the right data to homekit?

some more info to help answer this question: the dreams-integration does the job: it provides a map that updates in the view as a card on home assistant dashboard. Aside of it, it generates that camera entity that we are talking about it and that too shows up on the dashboard and updates when I go into the live stream. So I figured, why not bridge that camera into Homekit. All goes well when I go into homekit live stream, it stays buffered. 

@Tasshack the owner of dreame integration has tried to help us: homekit expects a certain type of video stream that homekit doesn't provide so they advise us to try to convert it before sending it to homekit (camera proxy for example). It didn't work so I turned to MQTT on advice on another user but that too fails for the reasons you mention above. 

thank you anyway for your time @sca075

**Reply by sca075** [reply]
@incline02 I'm not really familiar with HomeKit and as I wrote above this project is for Home Assistant and works with Valetudo (MQTT exposed) vacuums.

From what I understand in thread you had with @Tasshack the component he developed is exposing a camera that renders data for the valetudo card and yes I understand that in your environment (HomeKit) this do not render correctly the vacuum maps due to the format of the image.

The valetudo card is actually rendering the images from a PNG that has embedded a JSON with the data used from this component. If that json (zipped) is published to MQTT with a particular payload I could try to grab the data from there.. and basically add the code to support the integration of @Tasshack as I believe the goal of his project is to run as smoothly is possible on whatever architecture available in Home Assistant (I do have to warn you too that my component would not run on systems <PI4 8Gb although Poxmox is okay) as decoding and rendering the images is quite resource expensive as also commented in the Thread.

Feasibility wise.. I don't know if it's possible or not what I just wrote. Technically I don't know also if you can expose the card output from home assistant to homekit.. that so far I understand is actually your problem..

To me it looks like you really need to decode the maps, and render them (basically what this component do) and expose a camera to HomeKit.. but, again I don't know what payload we are looking for. 

So.. please clarify 
1) you have an Home Assistant instance and bridge the entities of it to HomeKit is that correct?
2) If I understand well your issue is that the current way is to expose the card? Or using the camera you get just a fix image? 
3) you need a camera entity in home assistant to expose the map data to HomeKit.. 
4) if you have an MQTT Explorer (add on for HA or stand alone) can you check for a payload map_data in your HA instance? It should be a camera entity.


If I understand it correctly there must be a camera entity in your MQTT or simply a payload similar to what I posted up. If you confirm that we can give it a try.

**Reply by incline02** [reply]
Apologies I only meant to give you more information in case you can think of alternative and send me on my way. I really appreciate your time.

To your questions:
1. Yes, HAOS on dedicated rasp and I used HomeKit Bridge integration to pull devices into homekit, I have several which work well so it's a working situation.
2. Here's the confusion: I'm not trying to expose the card and as you wrote, exposing anything would require a unique entity ID which dashboard elements don't have.
3. So I turn to the camera entity of the vacuum exposed by @tasshack's dreame integration: 
<img width="340" alt="image" src="https://github.com/user-attachments/assets/12013ca1-fa01-4a86-be40-728ac091c77e" />. When I used that alone and pull it into homekit, I get the infinite buffering indicating homekit can't decode. So I try to convert the stream via camera proxy and so I use proxy's entity instead, and same story. 
<img width="986" alt="image" src="https://github.com/user-attachments/assets/81a5d0ae-1976-4e12-a839-5a1c6e2d2d73" /> "Bot Current map" is the original; "camera proxy" is the proxy and "bot current map data" is to be ignored, it's for valetudo.

4. it's my first time with MQTT at all, I don't have an explorer. Will look it up. please standby

**Reply by incline02** [reply]
I've installed and went through the explorer. This is the result:

<img width="300" alt="image" src="https://github.com/user-attachments/assets/c5414478-43d5-40d3-9ac2-0200a936e894" />

(Running inside homeassistant) but I noticed you have a different first line showing an IP address

**Reply by sca075** [reply]
@incline02 had a look on the code of the integration you are using and I do not think you will found out the data in there. But so far I can see the Bot Current Map is actually already decoded and rendered as it should be... 

Probably there is some setting to tweet on your proxy.. but I think you got to investigate.

mmmm I can try one think only but this mean.. develop a different camera but before of all.. to let you try it I would need to create a dummy test camera. This could take a while.. mmmm.. the idea basically is the following:
- yaml config a camera that will use the default value.
- will render one or two dummy images to check if HomeKit is happy with those images.. if not sorry but can't help. 
if this solve the problem you have with the rendering (without proxy) probably we can use "bot current map data" and render your vacuum.. but this  work is not going to be publish as an official part of this project.. basically.. this would create a bridge for your bridge... it isn't a clean way to do this to be clear.

**Reply by incline02** [reply]
Well, I'm available for testing for the first part of your message but I agree a bridge of a bridge wouldn't be sustainable.

Whatever you choose, I'll make myself available.

**Reply by sca075** [reply]
@incline02 

Here we go this is a just a test camera that reproduces 3 different images at interval of 3 sec.
If this is rendered correctly in HomeKit we can give a try to decode the images from the integration you are currently using.

[test_camera.zip](https://github.com/user-attachments/files/18136810/test_camera.zip)

You will need to decompress the zip in the custom_components of Home Assistant and add to your configuration.yaml the following:
```yaml
camera:
  - platform: test_camera
    name: "Test Camera"
    images:
      - "custom_components/test_camera/test_png/test1.png"
      - "custom_components/test_camera/test_png/test2.png"
      - "custom_components/test_camera/test_png/test3.png"
    interval: 3
```

The 3 images was downloaded from wikipedia and are 3 desktop view.
If that works once again.. I can seriously think to get in contact with @Tasshack and try to found with him a solution.

**Reply by incline02** [reply]
just got our mail.

Following your instruction, I'll report as I'm done, please stand by.

**Reply by incline02** [reply]
unfortunately it doesn't work. Keeps buffering :/ (works as expected on HAOS dashboard)
<img width="1083" alt="image" src="https://github.com/user-attachments/assets/a6fdd982-5bc4-4e83-ac1b-6f719c376211" />

**Comment by Tasshack** [comment]
Please note that Dreame Vacuum integration has nothing to with MQTT or does not provide any MQTT connection or entity for it.

**Reply by incline02** [reply]
So back to square one on how to get the video stream for the entity decoded and encoded properly for homekit... We can close this topic then.

Like another user you discussed with, they and I tried everything but it doesn't work. I've got a netatmo camera > homeassistant > homekit bridge > homekit and it works so it really is just this stream from the bot entity.

**Reply by Tasshack** [reply]
You need to learn and master the ffmpeg usage for that.

**Comment by Tasshack** [comment]
You are the the wrong place since you dont have a rooted vacuum...

https://github.com/Tasshack/dreame-vacuum/issues/724

**Reply by incline02** [reply]
thanks for your replies

> You need to learn and master the ffmpeg usage for that.

I tried my hardest god knows I've programmed my Mac to wake up through an SSH command on a trigger from a motion sensor. But FFMPEG, I wasn't able to find a single exemple.

> https://github.com/Tasshack/dreame-vacuum/issues/724
I'm the last reply there :/

closing this thread

**Reply by Tasshack** [reply]
ffmpeg is a tool for converting any media format to any other media format so you cannot find examples for infinite possibilities, instead you need to read its documentation.

https://www.reddit.com/r/ffmpeg/comments/rw9vjz/convert_mjpeg_to_h264_livestream/
https://ffmpeg.org/ffmpeg-all.html

**Reply by incline02** [reply]
So I have tried to the best I could:
1. get the dynamic URL (token)
<img width="1560" alt="image" src="https://github.com/user-attachments/assets/4e997f2f-c4c9-4761-b37d-ff6eaa2558a1" />
2. Check if the URL is active by copy pasting the template result in a browser, works and I refreshed every time to make sure the token is valid
3. I copy pasted said URL in FFMPEG camera config
4. 
<img width="1173" alt="image" src="https://github.com/user-attachments/assets/6bff35f8-fbb5-46a9-9167-e57d624b8c99" />
5. I've also tried dynamic url (#'ed in the screenshot above
6. Finally I've tried extra_args and redone everything above.

None of the above works: the camera shows unavailable in the dashboard.

**Reply by incline02** [reply]
I think I'm gonna give up. It's getting excruciating.

I've tried 2 test url
1. https://www.earthcam.com/usa/louisiana/neworleans/bourbonstreet/?cam=bourbonstreet error: wrong header
2. http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/Sintel.mp4 barely works: freezes, buffers, wrong header
3. My own static urls extracted from template work in safari just like 1. and 2. but don't display via FFMPEG

none of the above work in got2RC or FFMPEG. got2rC is supposed to read anything but never worked. it's hair pulling

When I try my url, I don't even get header errors. I've tried tons of arguments, including -useragent, none would work.

People with the exact same issues never got answers, those with similar but different either don't use home assistant, or sometimes they're not sharing their configuration file, or they went another route which is much more complicated that just making ffmpeg work in config file or don't use the dreame L40. I'm in a pickle and building up anger, time to go outdoor

**Reply by sca075** [reply]
@incline02  

Actually Home Assistant has a kind of [native support for FFMPEG](https://www.home-assistant.io/integrations/camera.ffmpeg/).. it can also stream snapshots.. and of course.. there is to make a complete new implementation probably for lat it works with this kind of cameras we use for the maps...

I'be outdoor too have an I perfectly understand this can be a little frustrating... I'm anyhow the kind of person that do not give-up easily..  also knowing Apple quite well, nothing is 100% easy when using 3dt parity software and hardware with them product, but there isn't to disparate as "better to be frustrated because we try then doom as we never did try". 

Apparently there is a also an "ffmpeg-python" library that I'll study a little and can be useful too for the next level of implementation as there is a native I/O library that I'm willing to explore.. never the less.. I hope you can kindly understand how difficult could be to maintain a project like this or the one of @Tasshack natively designed for Home Assistant (not HomeKit).

**Reply by incline02** [reply]
alright, I've made progress:

The robot camera of the vacuum goes into a proxy camera which goes into go2rtc and success, I can see it in there as well as I've managed have a shell command update the token in go2rtc (thanks chat GPT).

Now ,go2rtc provides me with a dozen links to use into Proxycamera, Generic camera and/or FFMPEG but  bad news: none of those links work and this time, no error message other than "unvailable
